src/group/stabchain/builder/random.rs

```diff
@@ -0,0 +1,435 @@
+        g_iter
+            .flat_map(|g| {
+                subproduct_iter.iter().map(move |w| {
+                    let mut gw = g.clone();
+                    gw.extend(w.clone());
+                    gw
+                })
+            })
+            .collect()
+    }
+
+    /// Check if adding a new element modifies the current layer of the chain.
+    fn check_transversal_augmentation(&mut self, p: P) {
+        let mut record = self.chain[self.current_pos].clone();
+        //debug_assert!(record.gens.generators().contains(&p));
+        // If this element is already a generator, then we can exit
+        let mut to_check = VecDeque::from_iter(record.transversal.keys().cloned());
+        let mut new_transversal = HashMap::new();
+        while !to_check.is_empty() {
+            let orbit_element = to_check.pop_back().unwrap();
+            let new_image = self.action.apply(&p, orbit_element);
+
+            // If we haven't seen this element.
+            if !(record.transversal.contains_key(&new_image)
+                || new_transversal.contains_key(&new_image))
+            {
+                new_transversal.insert(new_image, p.inv());
+            }
+        }
+
+        // We now want to check all the newly added elements
+        let mut to_check = VecDeque::from_iter(new_transversal.keys().cloned());
+
+        // Update the record
+        record.transversal.extend(new_transversal);
+
+        // While we have orbit elements (and representatives to check)
+        while !to_check.is_empty() {
+            // Get the pair
+            let orbit_element = to_check.pop_back().unwrap();
+            // For each generator (and p)
+            for generator in std::iter::once(&p).chain(record.gens.generators()) {
+                let new_image = self.action.apply(&generator, orbit_element);
+                // If we haven't already seen the image
+                record.transversal.entry(new_image).or_insert_with(|| {
+                    // Update and ask to check the new image
+                    to_check.push_back(new_image);
+                    generator.inv()
+                });
+            }
+        }
+        // Update the generators adding p if it isn't already present.
+        if !record.gens.generators().contains(&p) {
+            record.gens =
+                Group::from_iter(std::iter::once(&p).chain(record.gens.generators()).cloned());
+        }
+        // Store the updated record in the chain
+        self.chain[self.current_pos] = record;
+    }
+
+    ///Check if the permutation augments the orbit at a level, resetting the position afterwards.
+    fn check_transversal_augmentation_at_level(&mut self, level: usize, p: P) {
+        let previous_pos = self.current_pos;
+        self.current_pos = level;
+        self.check_transversal_augmentation(p);
+        self.current_pos = previous_pos;
+    }
+
+    fn sgc(&mut self) {
+        println!("SGC {}/{}", self.current_pos, self.chain.len() - 1);
```

Comment:
Println!

Reply:
Ah, I was just so done with the code I forgot to remove them all. I've done that now.